File: nomarr/interfaces/api/web/navidrome.py
"""Navidrome integration endpoints for web UI."""

import logging
from typing import TYPE_CHECKING, Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from nomarr.helpers.exceptions import PlaylistQueryError
from nomarr.interfaces.api.auth import verify_session
from nomarr.interfaces.api.web.dependencies import get_navidrome_service

if TYPE_CHECKING:
    from nomarr.services.navidrome_service import NavidromeService

router = APIRouter(prefix="/navidrome", tags=["Navidrome"])


# ──────────────────────────────────────────────────────────────────────
# Request Models
# ──────────────────────────────────────────────────────────────────────


class PlaylistPreviewRequest(BaseModel):
    """Request model for playlist preview."""

    query: str = Field(..., min_length=1, description="Smart playlist query string")
    preview_limit: int = Field(10, ge=1, le=100, description="Number of sample tracks to return")


class PlaylistGenerateRequest(BaseModel):
    """Request model for playlist generation."""

    query: str = Field(..., min_length=1, description="Smart playlist query string")
    playlist_name: str = Field("Playlist", description="Name for the generated playlist")
    comment: str = Field("", description="Optional comment/description")
    sort: str | None = Field(None, description="Sort parameter (e.g., 'title', '-rating')")
    limit: int | None = Field(None, ge=1, le=10000, description="Maximum number of tracks")


# ──────────────────────────────────────────────────────────────────────
# Endpoints
# ──────────────────────────────────────────────────────────────────────


@router.get("/preview", dependencies=[Depends(verify_session)])
async def web_navidrome_preview(
    navidrome_service: "NavidromeService" = Depends(get_navidrome_service),
) -> dict[str, Any]:
    """Get preview of tags for Navidrome config generation (web UI proxy)."""
    try:
        stats = navidrome_service.preview_tag_stats()

        # Convert to list format for easier frontend consumption
        tag_list = []
        for tag_key, tag_stats in sorted(stats.items()):
            tag_list.append(
                {
                    "tag_key": tag_key,
                    "type": tag_stats["type"],
                    "is_multivalue": tag_stats["is_multivalue"],
                    "summary": tag_stats["summary"],
                    "total_count": tag_stats["total_count"],
                }
            )

        return {
            "tag_count": len(tag_list),
            "tags": tag_list,
        }

    except Exception as e:
        logging.exception("[Web API] Error generating Navidrome preview")
        raise HTTPException(status_code=500, detail=f"Error generating preview: {e}") from e


@router.get("/config", dependencies=[Depends(verify_session)])
async def web_navidrome_config(
    navidrome_service: "NavidromeService" = Depends(get_navidrome_service),
) -> dict[str, Any]:
    """Generate Navidrome TOML configuration (web UI proxy)."""
    try:
        toml_config = navidrome_service.generate_navidrome_config()

        return {
            "config": toml_config,
        }

    except Exception as e:
        logging.exception("[Web API] Error generating Navidrome config")
        raise HTTPException(status_code=500, detail=f"Error generating config: {e}") from e


@router.post("/playlists/preview", dependencies=[Depends(verify_session)])
async def web_navidrome_playlist_preview(
    request: PlaylistPreviewRequest,
    navidrome_service: "NavidromeService" = Depends(get_navidrome_service),
) -> dict[str, Any]:
    """Preview Smart Playlist query results."""
    try:
        # Call service method with validated request data
        result = navidrome_service.preview_playlist(query=request.query, preview_limit=request.preview_limit)

        # Convert result dataclass to dict for JSON response
        return {
            "total_count": result.total_count,
            "sample_tracks": result.sample_tracks,
            "query": result.query,
        }

    except PlaylistQueryError as e:
        raise HTTPException(status_code=400, detail=f"Invalid query: {e}") from e
    except HTTPException:
        raise
    except Exception as e:
        logging.exception("[Web API] Error previewing playlist")
        raise HTTPException(status_code=500, detail=f"Error previewing playlist: {e}") from e


@router.post("/playlists/generate", dependencies=[Depends(verify_session)])
async def web_navidrome_playlist_generate(
    request: PlaylistGenerateRequest,
    navidrome_service: "NavidromeService" = Depends(get_navidrome_service),
) -> dict[str, Any]:
    """Generate Navidrome Smart Playlist (.nsp) from query."""
    try:
        # Call service method with validated request data
        nsp_structure = navidrome_service.generate_playlist(
            query=request.query,
            playlist_name=request.playlist_name,
            comment=request.comment,
            sort=request.sort,
            limit=request.limit,
        )

        # Service returns the .nsp structure dict
        # Frontend expects: playlist_name, query, content, format
        # Convert dict to JSON string for "content" field
        import json

        return {
            "playlist_name": request.playlist_name,
            "query": request.query,
            "content": json.dumps(nsp_structure, indent=2),
            "format": "nsp",
        }

    except PlaylistQueryError as e:
        raise HTTPException(status_code=400, detail=f"Invalid query: {e}") from e
    except HTTPException:
        raise
    except Exception as e:
        logging.exception("[Web API] Error generating playlist")
        raise HTTPException(status_code=500, detail=f"Error generating playlist: {e}") from e


@router.get("/templates", dependencies=[Depends(verify_session)])
async def web_navidrome_templates_list() -> dict[str, Any]:
    """Get list of all available playlist templates."""
    try:
        from nomarr.helpers.navidrome_templates import get_template_summary

        templates = get_template_summary()
        return {"templates": templates, "total_count": len(templates)}

    except Exception as e:
        logging.exception("[Web API] Error listing templates")
        raise HTTPException(status_code=500, detail=f"Error listing templates: {e}") from e


@router.post("/templates", dependencies=[Depends(verify_session)])
async def web_navidrome_templates_generate() -> dict[str, Any]:
    """Generate all playlist templates as a batch."""
    try:
        from nomarr.helpers.navidrome_templates import generate_template_files

        templates = generate_template_files()
        return {"templates": templates, "total_count": len(templates)}

    except Exception as e:
        logging.exception("[Web API] Error generating templates")
        raise HTTPException(status_code=500, detail=f"Error generating templates: {e}") from e

File: frontend/src/shared/api/navidrome.ts
/**
 * Navidrome API functions.
 */

import { get, post } from "./client";

export interface NavidromeTag {
  tag_key: string;
  type: string;
  is_multivalue: boolean;
  summary: string;
  total_count: number;
  short_name: string;
  field_name: string;
  is_versioned: boolean;
}

interface NavidromePreviewRawResponse {
  stats: Record<string, {
    type: string;
    is_multivalue: boolean;
    summary: string;
    total_count: number;
    short_name: string;
    field_name: string;
    is_versioned: boolean;
  }>;
}

export interface NavidromePreviewResponse {
  tags: NavidromeTag[];
}

/**
 * Get preview of tags for Navidrome config.
 * Transforms backend stats dict into sorted tags array.
 */
export async function getPreview(): Promise<NavidromePreviewResponse> {
  const raw = await get<NavidromePreviewRawResponse>("/api/web/navidrome/preview");
  const tags = Object.entries(raw.stats)
    .map(([tag_key, stat]) => ({
      tag_key,
      type: stat.type,
      is_multivalue: stat.is_multivalue,
      summary: stat.summary,
      total_count: stat.total_count,
      short_name: stat.short_name,
      field_name: stat.field_name,
      is_versioned: stat.is_versioned,
    }))
    .sort((a, b) => a.short_name.localeCompare(b.short_name));
  return { tags };
}

export interface NavidromeConfigResponse {
  namespace: string;
  config: string;
}

/**
 * Generate Navidrome TOML configuration.
 */
export async function getConfig(): Promise<NavidromeConfigResponse> {
  return get("/api/web/navidrome/config");
}

export interface PlaylistPreviewResponse {
  total_count: number;
  sample_tracks: Array<Record<string, string>>;
  query: string;
}

/**
 * Preview Smart Playlist query results.
 */
export async function previewPlaylist(
  query: string,
  previewLimit = 10
): Promise<PlaylistPreviewResponse> {
  return post("/api/web/navidrome/playlists/preview", {
    query,
    preview_limit: previewLimit,
  });
}

export interface GeneratePlaylistParams {
  query: string;
  playlist_name: string;
  comment?: string;
  limit?: number;
  sort?: string;
}

export interface GeneratePlaylistResponse {
  playlist_name: string;
  query: string;
  content: string;
}

/**
 * Generate Navidrome Smart Playlist (.nsp).
 */
export async function generatePlaylist(
  params: GeneratePlaylistParams
): Promise<GeneratePlaylistResponse> {
  return post("/api/web/navidrome/playlists/generate", params);
}

export interface PlaylistTemplate {
  id: string;
  name: string;
  description: string;
  query: string;
  category?: string;
}

export interface GetTemplatesResponse {
  templates: PlaylistTemplate[];
  total_count: number;
}

/**
 * Get list of all available playlist templates.
 */
export async function getTemplates(): Promise<GetTemplatesResponse> {
  return get("/api/web/navidrome/templates");
}

export interface GeneratedTemplate {
  id: string;
  name: string;
  filename: string;
  success: boolean;
  error?: string;
}

export interface GenerateTemplatesResponse {
  templates: GeneratedTemplate[];
  total_count: number;
}

/**
 * Generate all playlist templates as a batch.
 */
export async function generateTemplates(): Promise<GenerateTemplatesResponse> {
  return post("/api/web/navidrome/templates");
}


// ── Tag Stats (for rules engine) ──

export interface TagStatEntry {
  key: string;
  type: "float" | "integer" | "string" | "unknown";
  is_multivalue: boolean;
  summary: string;
  total_count: number;
  short_name: string;
  field_name: string;
  is_versioned: boolean;
}

interface TagStatsRawResponse {
  stats: Record<string, {
    type: string;
    is_multivalue: boolean;
    summary: string;
    total_count: number;
    short_name: string;
    field_name: string;
    is_versioned: boolean;
  }>;
}

/**
 * Get tag metadata for the rules engine.
 * Transforms the backend stats dict into a sorted array of TagStatEntry.
 * Sorts by short_name for user-friendly display.
 */
export async function getTagStats(): Promise<TagStatEntry[]> {
  const raw = await get<TagStatsRawResponse>("/api/web/navidrome/preview");
  return Object.entries(raw.stats)
    .map(([key, stat]) => ({
      key,
      type: stat.type as TagStatEntry["type"],
      is_multivalue: stat.is_multivalue,
      summary: stat.summary,
      total_count: stat.total_count,
      short_name: stat.short_name,
      field_name: stat.field_name,
      is_versioned: stat.is_versioned,
    }))
    .sort((a, b) => a.short_name.localeCompare(b.short_name));
}


// ── Tag Values (for rules engine combobox) ──

interface TagValuesRawResponse {
  rel: string;
  values: string[];
}

/**
 * Get distinct values for a specific tag relationship.
 * Used by the rules engine combobox for autocomplete.
 */
export async function getTagValues(rel: string): Promise<string[]> {
  const raw = await get<TagValuesRawResponse>(
    `/api/web/navidrome/tag-values?rel=${encodeURIComponent(rel)}`,
  );
  return raw.values;
}